From: SeongJae Park <sj@kernel.org>
Date: Sat, 14 Feb 2026 10:47:45 -0800
Subject: [PATCH] mm/damon/core: respect online-updated min_nr_regions

DAMON API callers can update min_nr_regions online, using damon_call().
However, the upper limit of regions size, which is used for ensuring the
min_nr_regions, is updated only after ops update is done.  Check if
min_nr_regions is changed and update the upper limit.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index a03c0380e0fcc..ede3964d081cd 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -3295,6 +3295,8 @@ static int kdamond_fn(void *data)
 		unsigned long next_aggregation_sis = ctx->next_aggregation_sis;
 		unsigned long next_ops_update_sis = ctx->next_ops_update_sis;
 		unsigned long sample_interval = ctx->attrs.sample_interval;
+		/* cache min_nr_regions; it can be changed in damon_call() */
+		unsigned long old_min_nr_regions = ctx->attrs.min_nr_regions;
 
 		if (kdamond_wait_activation(ctx))
 			break;
@@ -3373,6 +3375,10 @@ static int kdamond_fn(void *data)
 				ctx->ops.update(ctx);
 			sz_limit = damon_region_sz_limit(ctx);
 		}
+
+		/* damon_call() callers could updated min_nr_regions */
+		if (ctx->attrs.min_nr_regions != old_min_nr_regions)
+			sz_limit = damon_region_sz_limit(ctx);
 	}
 done:
 	damon_destroy_targets(ctx);
-- 
2.47.3

